File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/HelpCommand.cs
using Atlasd.Daemon;
using Atlasd.Localization;
using System.Collections.Generic;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class HelpCommand : ChatCommand
    {
        public HelpCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            var hasAdmin = ChatCommand.HasAdmin(context.GameState);
            var topic = Arguments.Count > 0 ? Arguments[0] : string.Empty;
            if (!string.IsNullOrEmpty(topic)) Arguments.RemoveAt(0);
            var remarks = hasAdmin ? Resources.HelpCommandRemarksWithAdmin : Resources.HelpCommandRemarks;

            switch (topic.ToLower())
            {
                case "admin":
                    {
                        if (hasAdmin)
                        {
                            new AdminHelpCommand(RawBuffer, Arguments).Invoke(context); return;
                        }
                        break;
                    }
                case "advanced":
                    remarks = Resources.HelpCommandAdvancedRemarks; break;
                case "aliases":
                    remarks = Resources.HelpCommandAliasesRemarks; break;
                case "ban":
                    remarks = Resources.HelpCommandBanRemarks; break;
                case "channel":
                case "join":
                case "j":
                    remarks = Resources.HelpCommandJoinRemarks; break;
                case "commands":
                    remarks = Resources.HelpCommandCommandsRemarks; break;
                case "time":
                    remarks = Resources.HelpCommandTimeRemarks; break;
            }

            foreach (var kv in context.Environment)
            {
                remarks = remarks.Replace("{" + kv.Key + "}", kv.Value);
            }

            foreach (var line in remarks.Split(Battlenet.Common.NewLine))
            {
                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ChannelFlags, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/HelpCommand.vb
Imports AtlasV.Daemon
Imports AtlasV.Localization
Imports System.Collections.Generic

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class HelpCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim hasAdmin = varContext.GameState.HasAdmin()
            Dim topic = If(Arguments.Count > 0, Arguments(0), String.Empty)
            If Not String.IsNullOrEmpty(topic) Then Arguments.RemoveAt(0)
            Dim remarks = If(hasAdmin, Resources.HelpCommandRemarksWithAdmin, Resources.HelpCommandRemarks)

            Select Case topic.ToLower()
                Case "admin"

                    If hasAdmin Then
                        Call New AdminHelpCommand(RawBuffer, Arguments).Invoke(varContext)
                        Return
                    End If

                    Exit Select
                Case "advanced"
                    remarks = Resources.HelpCommandAdvancedRemarks
                Case "aliases"
                    remarks = Resources.HelpCommandAliasesRemarks
                Case "ban"
                    remarks = Resources.HelpCommandBanRemarks
                Case "channel", "join", "j"
                    remarks = Resources.HelpCommandJoinRemarks
                Case "commands"
                    remarks = Resources.HelpCommandCommandsRemarks
                Case "time"
                    remarks = Resources.HelpCommandTimeRemarks
            End Select

            For Each kv In varContext.Environment
                remarks = remarks.Replace("{" & kv.Key & "}", kv.Value)
            Next

            For Each line In remarks.Split(Battlenet.Common.NewLine)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
            Next
        End Sub
    End Class
End Namespace
